Add design doc for the network router feature (#1199)
nexodus-io/nexodus: docs/development/design/network-router.md

# Network Router

> Additional details and running code in [PR #981](https://github.com/nexodus-io/nexodus/pull/981)

## Summary

Add the ability for device's running nexd to access non-nexd devices and act as a router for external networks.

## Proposal

There are two supported use cases.

> 1. [Network Router with SNAT](#network-router-with-snat) Traffic from the Nexd mesh can reach external non-nexd nodes via source NAT without any configuration required.
> 2. [Network Router with SNAT disabled](#network-router-with-snat-disabled) Traffic from the Nexd mesh can reach external non-nexd nodes but require a route to be added for return traffic since SNAT is disabled.

Reviewer: ```suggestion
1. [Network Router with SNAT](#network-router-with-snat): Traffic from the Nexd mesh can reach external non-nexd nodes via source NAT without any configuration required.
2. [Network Router with SNAT disabled](#network-router-with-snat-disabled): Traffic from the Nexd mesh can reach external non-nexd nodes but external nodes require a route to be added for return traffic since SNAT is disabled.
```

> 3. With SNAT disabled, a user can initiate traffic from external, non-nexd devices to initiate traffic in to the nexodus mesh.

Reviewer: The section begins with saying there are 2 supported use cases, so I wouldn't make this a number 3 on the list. You could add it as another sentence with number 2, or just as a clarifying sentence afterwards like this.

```suggestion

With SNAT disabled, a user can initiate traffic from external, non-nexd devices to initiate traffic in to the nexodus mesh.
```

### Network Router with SNAT

This is a zero touch provisioning for the non-nexodus device since the edge nexd device will be source natting the traffic.

```shell

> nexd --username admin --password floofykittens router --child-prefix 192.168.100.0/24 --network-router  https://try.nexodus.127.0.0.1.nip.io

Reviewer: ```suggestion
nexd --username admin --password floofykittens --service-url https://try.nexodus.127.0.0.1.nip.io \
    router --child-prefix 192.168.100.0/24 --network-router  
```

```shell
# Linux
$ ip route | grep wg0
192.168.100.0/20 dev wg0
100.100.0.1 dev wg0
100.100.0.2 dev wg0
```

Here is an example with text diagramming a packet flow.

![no-alt-text](diagrams/network-router-snat.png)

The nftables configuration for this scenario look as follows:

```text
table inet nexodus-net-router {
    chain prerouting {
        type nat hook prerouting priority dstnat; policy accept;
    }

     chain postrouting {
        type nat hook postrouting priority srcnat; policy accept;
        oifname "eth0" counter packets 3 bytes 230 masquerade
    }

    chain forward {
        type filter hook forward priority filter; policy accept;
        oifname "eth0" ip daddr 192.168.100.0/24 counter packets 0 bytes 0 accept
    }
}
```

### Network Router with SNAT disabled

There are additional configurations required to make this scenario work:

- The non-nexd node needs to have a route, either static or from the underlying IGP pointing to the ingress interface on the nexd edge node.

> - It is not uncommon for there to be source/destination traffic checking enabled and blocked by default on cloud provider instances. For example, on EC2 one would need to disable [source / destination check](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-eni.html#modify-source-dest-check). The rest of the IP forwarding enablement is managed by the nexd agent.

Reviewer: ```suggestion
- It is common for there to be source/destination traffic checking enabled and blocked by default on cloud provider instances. For example, on EC2, one would need to disable [source / destination check](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-eni.html#modify-source-dest-check). The rest of the IP forwarding enablement is managed by the nexd agent.
```

# Network Router

> Additional details and running code in [PR #981](https://github.com/nexodus-io/nexodus/pull/981)

## Summary

Add the ability for device's running nexd to access non-nexd devices and act as a router for external networks.

## Proposal

There are two supported use cases.

1. Traffic from the Nexd mesh can reach external non-nexd nodes via source NAT without any configuration required.

> 2. Traffic from the Nexd mesh can reach external non-nexd nodes but require a configuration for return traffic since SNAT is disabled. This scenario will also allow external devices to initiate traffic to the mesh since the devices are required to have a route to an interface on the device connected to the physical network. With SNAT being disabled, the non-nexd device can also initiate traffic from outside the mesh going in.

Reviewer 2: Might be better to make #2 two different use cases even if they get implemented using the same strategy.

Author: You suggesting a separate doc or just putting it in a different section for NAT being disabled? Ty.

Reviewer 2: Just to add a third bullet to this doc.

### Network Router with SNAT

This is a zero touch provisioning for the non-nexodus device since the edge nexd device will be source natting the traffic.

```shell
nexd --username admin --password floofykittens router --child-prefix 10.180.0.0/20 --network-router  https://try.nexodus.127.0.0.1.nip.io

> ```

Reviewer 2: Does this command related to the following example diagram?

Author: Thanks, updated the prefix to match.

Here is an example with text diagramming a packet flow.

![no-alt-text](diagrams/network-router-snat.png)

> 

Reviewer 2: Maybe we should show that route table entries for 192.168.0.0 network are created by nexd in Node2
